### DISCUSSION TITLE
JTFS visuals

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#708] [Show and tell]
Exponential chirp; hopefully accurate, but not hard to redo.

First visuals are of two e-chirps going both ways (low to high, high to low), other two of just one (low to high).

### Joint wavelet slices, all pairs

All shaped `(7, 16)`, with per-gif color norms: `(0, 2e-5), (0, 1e-6), (0, 1e-6)` -- note: typo `inf -> -inf`

<img src="https://user-images.githubusercontent.com/16495490/113667872-74a93480-96c2-11eb-8d8f-7d2cb28a3d9d.gif" width="550">

<img src="https://user-images.githubusercontent.com/16495490/113667899-812d8d00-96c2-11eb-85f4-a0232be7636a.gif" width="380">

<img src="https://user-images.githubusercontent.com/16495490/113667951-94405d00-96c2-11eb-93c5-f8a5e094b681.gif" width="380">

### Joint wavelet slices (old)

<img src="https://user-images.githubusercontent.com/16495490/113455656-3f7bb880-941c-11eb-8de7-2aa5dde79c70.gif" width="450">

### All joint coefficients

Fully interactive, can dig inside

![image](https://user-images.githubusercontent.com/16495490/113455739-7baf1900-941c-11eb-8e8f-52e510044d57.png)

<hr>

### Code

<details>
  <summary>viz 1</summary>

```python
import matplotlib.pyplot as plt
from kymatio.numpy import TimeFrequencyScattering
from ssqueezepy import TestSignals
from ssqueezepy.visuals import imshow

#%%
def gif_jtfs(Scx, meta, norms=(1e-4, 1e-6, 1e-4), inf_token=-1, skip_spins=False):
    norms = [(0, n) for n in norms]
    n_spins = sum(s for s in meta['s'][1:] if s == 1)

    i = 1  # skip first-order

    # `psi_t * psi_f` first
    while inf_token not in meta['n'][i + n_spins]:
        if not skip_spins:
            _viz_spins(Scx, meta, i, norms[0], n_spins)
        i += 1
    i = (i - 1) * 2 + 1

    # `psi_t * phi_f`
    while meta['n'][i][1] == inf_token:
        _viz_simple(Scx, meta, i, norms[1])
        i += 1

    # `phi_t * psi_f`
    while meta['n'][i][0] == inf_token:
        _viz_simple(Scx, meta, i, norms[2])
        i += 1
        if i >= len(Scx):
            break


def _title(meta, i, spin):
    txt = r"$|\Psi_{%s, %s, %s} \star X| \star \phi_T \star \phi_F$"
    mu, l = [int(n) if (float(n).is_integer() and n >= 0) else '\infty'
             for n in meta['n'][i]]

    return (txt % (mu, l, spin), {'fontsize': 20})

def _viz_spins(Scx, meta, i, norm, n_spins):
    i0, i1 = i, i + n_spins

    fig, axes = plt.subplots(1, 2, figsize=(15, 7))
    kw = dict(abs=1, ticks=0, show=0, norm=norm)
    imshow(Scx[i0]['coef'], ax=axes[0], **kw, title=_title(meta, i, '+1'))
    imshow(Scx[i1]['coef'], ax=axes[1], **kw, title=_title(meta, i, '-1'))
    plt.subplots_adjust(wspace=0.01)
    plt.show()

def _viz_simple(Scx, meta, i, norm):
    imshow(Scx[i]['coef'], abs=1, ticks=0, show=1, norm=norm, w=.8, h=.5,
           title=_title(meta, i, '0'))

#%%###########################################################################
T = 8192
J = 9
Q = 16
scattering = TimeFrequencyScattering(J, T, Q, J_fr=4, Q_fr=2, out_type='list')
tsigs = TestSignals(N=T)

#%%
x  = tsigs.make_signals('echirp') / 2
x += x[::-1]
Scx = scattering(x)

#%%
gif_jtfs(Scx, scattering.meta())
```
</details>

<details>
  <summary>viz 2</summary>

```python
from kymatio.numpy import TimeFrequencyScattering
from ssqueezepy import TestSignals
from ssqueezepy.visuals import imshow

#%%
T = 4096*2
J = 6
Q = 12
scattering = TimeFrequencyScattering(J, T, Q)

#%%
x = TestSignals(N=T).echirp()[0]
Scx = scattering(x)

#%%
txt = r"$|\Psi_{%s, %s, +1} \star X| \star \phi_T \star \phi_F$"

for S in Scx[1:]:  # skip first-order
    imshow(S['coef'], abs=1, norm=(0, 1e-4), ticks=0,
           title=(txt % S['n'], {'fontsize': 12}))
# https://ezgif.com/maker
```
</details>

<details>
  <summary>viz 3</summary>

```python
# repeat viz 1 to get `Scx`
import numpy as np
import plotly.graph_objs as go
from plotly.offline import plot  # for IDE (e.g. Spyder use)

#%%# Prepare to plot ##
arr = np.vstack([S['coef'][None] for S in Scx[1:]])
a, b, c = arr.shape
l = 1
X, Y, Z = np.mgrid[-l:l:a*1j, -l:l:b*1j, -l:l:c*1j]

#%%# Plot in browser #########################################################
fig = go.Figure(data=go.Volume(
    x=X.flatten(),
    y=Y.flatten(),
    z=Z.flatten(),
    value=arr.flatten(),
    opacity=0.25, # small to see through surfaces
    surface_count=12, # larger -> better volume rendering
    colorscale='jet',
    ))
plot(fig, auto_open=True)
```
</details>

**Comment by lostanlen** [comment]
Awesome!!

@changhongw thoughts?

**Reply by OverLordGoldDragon** [reply]
Added pairwise; `Scx` no longer needs to be list

**Reply by changhongw** [reply]
This is fantastic!

One question: for visualization 1, why the figures for spin +1 and -1 appear semmetrically? For an exponential chirp with a specific direction, shouldn't the energy concentrate on one of the spins?

**Reply by OverLordGoldDragon** [reply]
@changhongw Right, that one's different from other two, two chirps going both ways - edited for clarity.

**Comment by lostanlen** [comment]
Is there a way we could display the joint time-frequency wavelets? no averaging, no complex modulus, just the real part of `Psi(t,f)` where `Psi` is either `psi_t * psi_f`, `psi_t * phi_f`, or `phi_t * psi_f`
In `scattering.m` i achieved this by passing a 2-D Dirac (a single nonzero pixel at the center of the time frequency domain) to the JTFS wavelet transform.

The output would look something like Fig 1 of : https://www.lostanlen.com/wp-content/uploads/2019/06/lostanlen2019dafx.pdf
(note that there was no `phi_t * psi_f` in that figure because the `S1` was averaged in time but unaveraged in frequency)

in this way we could demonstrate that JTFS in Kymatio is similar to spectro-temporal receptive fields (STRF), except much faster
Another point of comparison is the gabor filterbank features (GBFB) of Schädler https://medi.uni-oldenburg.de/members/bernd/papers/JASA_2012_Schaedler.pdf

@OverLordGoldDragon i think Figures 4 and 6 will interest you

**Reply by OverLordGoldDragon** [reply]
Wavelet visuals would be nice, may look into, but ideally we'll plot those directly rather than relying on input to work properly since all sorts of implementation-related artifacts could manifest.

It'd help to automate the gif generation process; there's `ffmpeg` but a pain to get working on Windows.

**Reply by lostanlen** [reply]
>  all sorts of implementation-related artifacts could manifest

i'd argue that it's a good thing. it's important to see the effect of discretization, boundary conditions, etc. 
at the same time, you're right that the simplest approach is to just call the filterbank constructors and compute interference products in 2D Fourier domain without any subsampling 
(this is how STRF is implemented by the way)

Perhaps relevant: https://github.com/EtienneTho/strf-like-model/blob/main/auditory.py

**Reply by OverLordGoldDragon** [reply]
>   it's important to see the effect of discretization, boundary conditions, etc.

Definitely, so we'll have a complete picture with direct wavelet plots as "ideal" reference

**Reply by lostanlen** [reply]
PS: note that STRF uses `Q2=2` and `Q_fr=2`. This is rooted in domain-specific knowledge about auditory neuroscience. It would be great if we could reproduce the same output as STRF yet with the multirate approach you developed 
(also STRF doesn't have a GPU implementation)

**Reply by changhongw** [reply]
Agree. @lostanlen Can we have a variable Q2 that can be defined by the user instead of the current Q2=1 only, for joint scattering? Whether the spectro-temporal modulation can be clearly captured depends highly on the match of Q2 and Q_fr.

**Reply by OverLordGoldDragon** [reply]
What's Q2?

@changhongw The default we had was `Q_fr=1`, which is now [user-exposed](https://github.com/kymatio/kymatio/blob/aaefe31e2c38c8082ae05ddf0064d26b38d0d410/kymatio/scattering1d/frontend/numpy_frontend.py#L88)

**Reply by lostanlen** [reply]
Q2 is the number of wavelets per octave of the temporal filterbank at the second layer

**Reply by OverLordGoldDragon** [reply]
@lostanlen Ah, duh. Agreed with @changhongw. Then should we also default `Q2` to `Q_fr`, and vice versa, if the user doesn't specify both?

**Reply by changhongw** [reply]
Default Q2 to 1 but can be user defined? For example, it can be defined by the user as Q=[16, 2] or Q=16. For the first case, Q1=16 and Q2=2; for the second case, Q1=16, and Q2=1 by default.

**Reply by OverLordGoldDragon** [reply]
@changhongw Right, I meant if user does `Q2=2` but nothing for `Q_fr`, then we default `Q_fr=2`? Likewise if user passes `Q_fr=3` but nothing for `Q2`, we default `Q2=3`

**Reply by lostanlen** [reply]
i'm fine with any default as long as it's documented. regarding Q1, i'd argue that it's important to not set any default so that the user knows it's their responsability to choose a number of wavelets per octave at the first layer.
What would make the most sense for replicating STRF in auditory neurophysiology would be Q2=2, Q_fr=2.

~@changhongw any preference re: default values of Q?~ thanks for chiming in

**Reply by changhongw** [reply]
Agree. Q2=2 and Q_fr=2 may be a good start. I would suggest to set the default values for Q2 and Q_fr separately, rather than coupling them by Q2=Q_fr. If the signal contains high temporal modulation oscillations while that of the frequnetial modulation is pretty low. When the user sets a large Q2=8, in this case, Q_fr=8 may be not a good value.

**Reply by OverLordGoldDragon** [reply]
@changhongw You suggest we change defaults to `Q2=2` and `Q_fr=2` for JTFS?

**Reply by changhongw** [reply]
Yes. Will this introduce a problem when merging into Kymatio later? Currently Kymatio use Q2=1 as default for time scattering. How about follow this, i.e. default Q2=1? For example, modify the function calibrate_scattering_filters in filter_bank.py into:

<details>

```python
def calibrate_scattering_filters(J, Q, r_psi=math.sqrt(0.5), sigma0=0.1,
                                 alpha=5.):

    sigma_low = sigma0 / math.pow(2, J)  # width of the low pass
    if isinstance(Q, int):
        xi1, sigma1, j1 = compute_params_filterbank(sigma_low, Q, r_psi=r_psi,
                                                alpha=alpha)
        xi2, sigma2, j2 = compute_params_filterbank(sigma_low, 1, r_psi=r_psi,
                                            alpha=alpha)
    elif Q[0] < 1:
        raise ValueError('Q should always be >= 1, got {}'.format(Q))
    else:
        xi1, sigma1, j1 = compute_params_filterbank(sigma_low, Q[0], r_psi=r_psi,
                                                alpha=alpha)
        xi2, sigma2, j2 = compute_params_filterbank(sigma_low, Q[1], r_psi=r_psi,
                                                alpha=alpha)

    return sigma_low, xi1, sigma1, j1, xi2, sigma2, j2
```
</details>

**Reply by OverLordGoldDragon** [reply]
@changhongw Yeah, it's hard-coded. This should probably be a separate PR, but I'll add it to JTFS temporarily to see it if JTFS needs any changing, then both can merge.

**Comment by OverLordGoldDragon** [comment]
No exact fit so I'll put here - docs supplement for `toolkit.pack_coeffs_jtfs`.

`n1_fr` is *tensor index* along frequential scattering slices, `n2` along temporal. `3` i.e. `structure == 3`.

<img src="https://user-images.githubusercontent.com/16495490/141265017-40b5fcac-9fb9-408a-901c-3c0d24d7727c.png" width="400">

<img src="https://user-images.githubusercontent.com/16495490/141265751-45b710b8-6c4b-4386-bc4d-5fcc688f0e82.png" width="480">
